### PR TITLE
Make `NonZero` constructors generic.

### PR DESCRIPTION
**Pull request by reitermarkus**
This makes `NonZero` constructors generic, so that `NonZero::new` can be used without turbofish syntax.

Tracking issue: https://github.com/rust-lang/rust/issues/120257

~~I cannot figure out how to make this work with `const` traits. Not sure if I'm using it wrong or whether there's a bug:~~

```rust
101 |         if n == T::ZERO {
    |            ^^^^^^^^^^^^ expected `host`, found `true`
    |
    = note: expected constant `host`
               found constant `true`
```

r? @dtolnay

### CODE REVIEW

File: library/core/src/num/nonzero.rs [hunk 1]
```diff
@@ -83,6 +88,79 @@ impl_zeroable_primitive!(NonZeroIsize(isize));
...
+    #[rustc_const_stable(feature = "const_nonzero_int_methods", since = "1.47.0")]
+    #[must_use]
+    #[inline]
+    pub const fn new(n: T) -> Option<Self> {
+        if n == T::ZERO {
+            return None;
+        }
+
+        // SAFETY: We just checked that there's no `0`.
+        Some(unsafe { Self(n) })
```
Comment (dtolnay) [inline library/core/src/num/nonzero.rs:106]: This should fix the const traits issue.

```suggestion
        unsafe { crate::mem::transmute_copy(&n) }
```

**Review by dtolnay — CHANGES_REQUESTED** [review]

**Comment by Nilstrieb** [comment]
@rust-lang/project-const-traits looks like a bug. Should this be relying on const traits in stable code anyways, are they ready and bug free enough for that?

**Comment by oli-obk** [comment]
> @rust-lang/project-const-traits looks like a bug.

just a diagnostics bug, the error was correct.



> Should this be relying on const traits in stable code anyways, are they ready and bug free enough for that?

nope, should definitely not use them in stable code.

File: library/core/src/num/nonzero.rs (hunk 1 above)
Comment (oli-obk) [reply library/core/src/num/nonzero.rs:106]:
```rust
match n {
    T::ZERO => None,
    _ => Some(unsafe { self(n) }),
}
```

may also work without requiring a PartialEq impl

**Review by oli-obk — APPROVED** [review]
impl lgtm. the main PR message needs updating.

cc #120257 for backlinking

**Review by dtolnay — APPROVED** [review]
Thanks!

File: library/core/src/num/nonzero.rs [hunk 2]
```diff
@@ -83,6 +83,90 @@ impl_zeroable_primitive!(NonZeroIsize(isize));
...
 pub struct NonZero<T: ZeroablePrimitive>(T);
 
+impl<T> NonZero<T>
+where
+    T: ZeroablePrimitive,
+{
+    /// Creates a non-zero if the given value is not zero.
+    #[stable(feature = "nonzero", since = "1.28.0")]
+    #[rustc_const_stable(feature = "const_nonzero_int_methods", since = "1.47.0")]
+    #[rustc_allow_const_fn_unstable(const_refs_to_cell)]
```
Comment (fee1-dead) [inline library/core/src/num/nonzero.rs:93]: Uh, is this needed?

Comment (oli-obk) [reply library/core/src/num/nonzero.rs:93]: for the `transmute_copy` taking a reference to a generic parameter. Unfortunately we can't use `transmute`, even though we know the size is the same, rustc doesn't.

We could teach the transmute checks about `rustc_nonnull_optimization_guaranteed` and use `transmute` again.

**Comment by matthiaskrgr** [comment]
could this failure be tired to this pr?
https://github.com/rust-lang/rust/pull/120667#issuecomment-1926668781

File: library/core/src/num/nonzero.rs [hunk 3]
```diff
@@ -83,6 +84,88 @@ impl_zeroable_primitive!(NonZeroIsize(isize));
...
+    #[must_use]
+    #[inline]
+    pub const fn new(n: T) -> Option<Self> {
+        // SAFETY: Memory layout optimization guarantees that `Option<NonZero<T>>` has
+        //         the same layout and size as `T`, with `0` representing `None`.
+        unsafe { ptr::read(ptr::addr_of!(n).cast()) }
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply library/core/src/num/nonzero.rs:100]: it used to be. I assume this was changed to be able to remove the `#[rustc_allow_const_fn_unstable(const_refs_to_cell)]`, but then that wasn't removed 😆

[the remark this answers is not in the document]
Comment (reitermarkus) [reply library/core/src/num/nonzero.rs:100]: No, it was removed to fix the failing `niche-filling.rs` test, caused by the `assert!` call in `transmute_copy`. Maybe `transmute_unchecked` is what I was looking for.

[the remark this answers is not in the document]
Comment (reitermarkus) [reply library/core/src/num/nonzero.rs:100]: I tried removing `#[rustc_allow_const_fn_unstable(const_refs_to_cell)]` as well, but it seems to be used by `addr_of!`.

**Comment by Kobzol** [comment]
Regression mostly resolved by #120809.

@rustbot label: +perf-regression-triaged

**Pull request merged**